python/lsst/pipe/tasks/processBrightStars.py

```diff
@@ -0,0 +1,414 @@
+            Positions of each object's center (as obtained from the refCat),
+            in pixels.
+
+        Returns
+        -------
+        warpedStars : `list` [`afwImage.maskedImage.maskedImage.MaskedImage`]
+        """
+        # warping control; only contains shiftingALg provided in config
+        warpCont = afwMath.WarpingControl(self.config.warpingKernelName)
+        # Compare model to star stamp sizes
+        bufferPix = (self.modelStampSize[0] - self.config.stampSize[0],
+                     self.modelStampSize[1] - self.config.stampSize[1])
+        # Initialize detector instance (note all stars were extracted from an
+        # exposure from the same detector)
+        det = stamps[0].getDetector()
+        # Define correction for optical distortions
+        pixToTan = det.getTransform(cg.PIXELS, cg.TAN_PIXELS)
+        # Array of all possible rotations for detector orientation:
+        possibleRots = np.array([k*np.pi/2 for k in range(4)])
+        # determine how many, if any, rotations are required
+        yaw = det.getOrientation().getYaw()
+        nb90Rots = np.argmin(np.abs(possibleRots - float(yaw)))
+
+        # apply transformation to each star
+        warpedStars = []
+        for star, cent in zip(stamps, pixCenters):
+            # (re)create empty destination image
+            destImage = afwImage.MaskedImageF(*self.modelStampSize)
+            bottomLeft = geom.Point2D(star.getImage().getXY0())
+            newBottomLeft = pixToTan.applyForward(bottomLeft)
+            newBottomLeft.setX(newBottomLeft.getX() - bufferPix[0]/2)
+            newBottomLeft.setY(newBottomLeft.getY() - bufferPix[1]/2)
+            # Convert to int
+            newBottomLeft = geom.Point2I(newBottomLeft)
+            # Set origin
+            destImage.setXY0(newBottomLeft)
+
+            # Define linear shifting to recenter stamps
+            newCenter = pixToTan.applyForward(cent)  # center of warped star
+            shift = self.modelCenter[0] + newBottomLeft[0] - newCenter[0],\
+                self.modelCenter[1] + newBottomLeft[1] - newCenter[1]
+            affineShift = geom.AffineTransform(shift)
+            shiftTransform = tFactory.makeTransform(affineShift)
+
+            # Define full transform (warp and shift)
+            starWarper = pixToTan.then(shiftTransform)
+
+            # Apply it
+            goodPix = afwMath.warpImage(destImage, star.getMaskedImage(),
+                                        starWarper, warpCont)
+            if not goodPix:
+                self.log.debug("Warping of a star failed: no good pixel in output")
+
+            # Arbitrarily set origin of shifted star to 0
+            destImage.setXY0(0, 0)
+
+            # Apply rotation if apropriate
+            if nb90Rots:
+                destImage = afwMath.rotateImageBy90(destImage, nb90Rots)
+            warpedStars += [destImage.clone()]
+        return warpedStars
+
+    def computeAnnularFlux(self, image):
```

Comment:
I have given this some thought, and I still think I would rather keep it a `ProcessBrightStarsTask` method.

Firstly, note that we also make extensive use of other fields from the config, that define exactly what statistic is used to compute the flux, as well as which mask planes should be ignored.

I think this reveals that this computation is very much part of the _processing_ of bright stars. My thinking on what `BrightStarStamps` should be is that these have already been "processed." They contain all the information necessary to revert that processing (or make use of it, eg to scale the eventual model before subtraction), but that processing could have been done in a number of ways.

Keeping things this way - with the computation of the annularFlux and the normalization of stamps in the Task rather than as a method of the `BrightStarStamps` - guarantees that `BrightStarStamps` are a set of normalized bright star stamps, and that the annular fluxes contained have all been computed in the same way. 

Does that make sense?

I did shuffle things around a bit, however. As you suggested, the method now measures _and_ normalizes stamps in place. The loop over stamps is done within the method, as is already the case with `warpStamps`. This actually avoids recomputing the SpanSet every single time, since it is the exact same for each stamp. (I was hoping I could get away with handling mask planes just the once as well, but I feel like this might not be perfectly safe if the `maskPlaneDict` changes from one stamp to another, so I still redo it on a per-stamp basis.)
I feel like this also alleviates the awkwardness in `run` somewhat.

Reply:
I would only refute your claim that this behavior guarantees that `BrightStarStamps` are normalized set. They are normalized because of the implementation details here. But nothing it stopping someone from inserting a non-normalized one into a `BrightStarStamps` object when not using this way of creating them (or appending one later). In contrast if this method was run by default on those objects that would ensure the objects were normalized unless someone explicitly opted out (`normalize=False` or something). As for config parameters, you would only need to pass in an optional `statsFlag` object and `statsConfig` object, that would have default implementations of some kind.

I think you are putting a lot of weight on "how I implemented it to work here" being equal to "This is the behavior of the classes" when you need to consider someone might use one w/o the other, or later used some subclass.

This is not to say I am unhappy with how this is implemented, nor do I think this is not good enough to merge. I just want to bring up corners for you to consider.

Reply:
In the end, I went ahead and changed it as you suggested.

This leads to changes in meas_algorithm, and the following PR: https://github.com/lsst/meas_algorithms/pull/220